src/object/cli_obj.c

```diff
@@ -2627,3 +2662,369 @@ dc_obj_query_key(tse_task_t *api_task)
+
+		rc = daos_iod_copy(&iods_out[count], &iods_in[i]);
+		if (rc != 0)
+			goto out;
+
+		count++;
+		iods_in[i].iod_recxs = NULL;
+		iods_in[i].iod_csums = NULL;
+		iods_in[i].iod_eprs = NULL;
+	}
+
+	/* All data are inline, not need further check. */
+	if (count == 0)
+		goto out;
+
+	D_ALLOC_ARRAY(sgls, count);
+	if (sgls == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	D_ALLOC_ARRAY(iovs, count);
+	if (iovs == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	size = 0;
+	for (i = 0; i < count; i++) {
+		iovs[i].iov_buf_len = daos_iods_len(&iods_out[i], 1);
+		D_ASSERT(iovs[i].iov_buf_len != DAOS_REC_ANY);
+		D_ASSERT(iovs[i].iov_buf_len != 0);
+
+		size += iovs[i].iov_buf_len;
+	}
+
+	D_ASSERT(unpack_arg->arg != NULL);
+	D_ASSERT(unpack_arg->arg->fetch_buf == NULL);
+	D_ASSERT(unpack_arg->arg->data_size == 0);
+
+	D_ALLOC(unpack_arg->arg->fetch_buf, size);
+	if (unpack_arg->arg->fetch_buf == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	unpack_arg->arg->data_size = size;
+
+	size = 0;
+	for (i = 0; i < count; i++) {
+		iovs[i].iov_buf = &unpack_arg->arg->fetch_buf[size];
+		size += iovs[i].iov_buf_len;
+
+		sgls[i].sg_nr = 1;
+		sgls[i].sg_nr_out = 1;
+		sgls[i].sg_iovs = &iovs[i];
+	}
+
+	coh = dc_obj_hdl2cont_hdl(unpack_arg->oh);
+	if (daos_handle_is_inval(coh))
+		D_GOTO(out, rc = -DER_NO_HDL);
+
+	rc = dc_tx_local_open(coh, epoch, &th);
+	if (rc != 0)
+		goto out;
+
+	rc = dc_obj_fetch_sepc_wrap(unpack_arg->oh, th, DIOF_TO_SPEC_SHARD,
+				    unpack_arg->shard, &io->ui_dkey, count,
+				    iods_out, sgls);
+	if (rc != 0) {
+		D_ERROR(DF_OID" (shard %u) failed to fetch data: %d\n",
+			DP_OID(unpack_arg->oid), unpack_arg->shard, rc);
+		D_GOTO(out, rc);
+	}
+
+	for (i = 0; i < count; i++) {
+		if (iovs[i].iov_len != iovs[i].iov_buf_len) {
+			/* XXX: Claimed non-zero length, but fetched nothing,
+			 *	then either some modification DTX (with some
+			 *	lower epoch than the given one) is committed
+			 *	by race between enumerate and fetch, or some
+			 *	old epochs and related records are removed by
+			 *	vos aggregation.
+			 *
+			 *	Currently, we assume that the DTX race will
+			 *	not happen via given a stable @epoch for the
+			 *	consistency verification. So it will be the
+			 *	latter case. Let's return -DER_AGAIN to the
+			 *	caller and retry the list and fetch from the
+			 *	beginning of the shard because the cursor in
+			 *	the anchor may become stale because of rare
+			 *	vos aggregation.
+			 */
+			D_DEBUG(DB_TRACE,
+				DF_OID" (shard %u) claim %ld bytes, "
+				"but fetched %ld bytes\n",
+				DP_OID(unpack_arg->oid), unpack_arg->shard,
+				iovs[i].iov_buf_len, iovs[i].iov_len);
+			D_GOTO(out, rc = -DER_AGAIN);
+		}
+	}
+
+out:
+	if (iods_out != NULL)
+		daos_iods_free(iods_out, count, true);
+
+	if (sgls != NULL)
+		D_FREE(sgls);
+
+	if (iovs != NULL)
+		D_FREE(iovs);
+
+	if (!daos_handle_is_inval(th))
+		dc_tx_local_close(th);
+
+	return rc;
+}
+
+static int
+dc_obj_verify_one_rdg(struct dc_object *obj, unsigned int start,
```

Comment:
I really think the "_one_" is unnecessary and we can remove it from these functions and data structures. RDB == redundancy group, it is clear enough.